### PR TITLE
HHH-16040 AssertionError for @JoinColumn with char[] property referencing a String property

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-16040
https://hibernate.atlassian.net/browse/HHH-15916

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/TreatPathTest.java [hunk 1]
```diff
@@ -129,13 +131,17 @@ public void testTreatAnyValue(EntityManagerFactoryScope scope) {
...
+	}
+
+	private void testCriteriaTreat(EntityManager entityManager, String property, Object value, boolean useIn) {
+		final CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+		final CriteriaQuery<Linkage> criteria = cb.createQuery( Linkage.class );
+		final Root<Linkage> root = criteria.from( Linkage.class );
+		final Path<LocalTerm> asLocalTerm = cb.treat( root.get( "term" ), LocalTerm.class );
+		final Predicate predicate;
+		if ( useIn ) {
+			predicate = cb.literal( value ).in( asLocalTerm.get( property ) );
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/TreatPathTest.java:144]: This refactoring kind of seems wrong. You switched from `asLocalTerm.get( property ).in( value )` to `cb.literal( value ).in( asLocalTerm.get( property ) )`, which is reverse AFAIU.

**Review by beikov — COMMENTED** [review]
Other than what I commented, looks good to me!

File: hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/TreatPathTest.java (hunk 1 above)
Comment (mbladel) [reply hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/TreatPathTest.java:144]: That was on purpose, I think the way I wrote the test before was wrong. This case with plural values is a bit flaky anyway - I thought about disabling it but for the purpose of checking treat itself it was enough to expect an `UsupportedOperationException`.

Comment (mbladel) [reply hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/TreatPathTest.java]: Fixed the test now.

**Review by beikov — COMMENTED** [review]
It seems `CharArrayToStringInEmbeddedJoinColumnOrFormulaTest` fails on Sybase for some reason.

**Comment by mbladel** [comment]
> It seems `CharArrayToStringInEmbeddedJoinColumnOrFormulaTest` fails on Sybase for some reason.

Yep, it was caused by the `trim` function in the formula. I just removed it since it's not relevant for the purpose of the test.

**Review by beikov — APPROVED** [review]
LGTM

**Pull request merged**